### ISSUE TITLE
bdshemu: support for SIMD instructions

### ISSUE THREAD

**Issue by JustasMasiulis** [bitdefender/bddisasm#41]
Are there any plans to add support for SSE and potentially even AVX instructions?

If not could I get some pointers into how I could attempt to implement this? While the code seems quite straightforward at the same time it's looks like a lot of hardcoded operand indices that I could very easily make a mistake with.

On a side note, would there be any possibility of a more stripped down emulation API? I am writing a debugger where I want to use bdshemu as a generic code emulator and for that I had to remove all the buffers in the emulator context and replace them with the memory access callback. I feel like this could be useful to a much wider audience of people because it would be more flexible and usable not only for shellcode emulation.

**Comment by vlutas** [comment]
Hello and thank you for your interest in this project! :)

There are no particular plans of adding support for SSE/AVX instructions - the Shellcode Emulator is extended with new instructions only when needed to support additional shellcode capabilities. It is not intended to be a full instruction emulator, this is why the emulation API is very specific to shellcodes (including the fact that it normally doesn't allow any external memory accesses).

Please note that since the emulator is highly specific to shellcodes, it does not contain full instruction emulation (for example, checking access rights, alignment checks, etc.) which are normally important for instruction emulation. If you wish to use it as a general purpose instruction emulation, it must be extended with such features.

As for operand indices - they are exactly as defined by the Intel SDM. Where additional, implicit operands are involved, you can consult the instruction tables (isagenerator/instructions/table_*.dat) to see the order of the implicit operands.

Like always, pull-requests are welcome. If you wish to add support for new instructions, or new API, contributions are very much apreciated! :)

**Comment by ianichitei** [comment]
I'd like to add that a more general purpose emulator built on top of `bddisasm` is used by the Napoca hypervisor: check [NdEmulateInstruction](https://github.com/napocahv/napoca/blob/HEAD/napoca/kernel/emu.c#L7675) - this has support for more instructions and also handles the exceptions that may be triggered by the instructions it emulates, which is probably something that you want in your emulator, although I'm not entirely sure I completely understand your use case.

**Comment by JustasMasiulis** [comment]
Thanks for the reply.

> it does not contain full instruction emulation

> exceptions that may be triggered by the instructions it emulates, which is probably something that you want in your emulator, although I'm not entirely sure I completely understand your use case.

I understand that and I'm perfectly fine with it. My use case is very crude and doesn't really require 100% accuracy. Currently all I want to use the emulator for is tracing control flow paths to highlight them in debugger UI.

The main reason I chose bddisasm + bdshemu is because they are lightweight and come in a compact package. This matters to me because the UI/frontend is made with web technologies. Disassembler + emulator + custom instruction formatting weighs in at less than 500kb of web assembly uncompressed, whereas something like [unicorn.js](https://alexaltea.github.io/unicorn.js/) is almost 20 megabytes alone.

> I'd like to add that a more general purpose emulator built on top of bddisasm is used by the Napoca hypervisor

Haven't seen that and it looks great, however it would again require quite a bit of modification to rip it out of the hypervisor.

> Like always, pull-requests are welcome. If you wish to add support for new instructions, or new API, contributions are very much apreciated! :)

Thanks for all the information! After my project moves forward a bit with the more essential features/basics I'll probably come back to this and try extending functionality a bit. (although knowing myself this might take a while ™️)

**Comment by ianichitei** [comment]
> This matters to me because the UI/frontend is made with web technologies. Disassembler + emulator + custom instruction formatting weighs in at less than 500kb of web assembly uncompressed.

This sounds intriguing. Have you been able to compile bdshemu with Emscripten? I'm curious what changes were necessary for that to work.

> Haven't seen that and it looks great, however it would again require quite a bit of modification to rip it out of the hypervisor.

It's easier just to use it as an inspiration source. Using bdshemu as a starting point is the easiest way, but it would have to be a different library/project since bdshemu is not meant to be a general purpose emulator.

**Comment by vlutas** [comment]
If your use-case doesn't imply something more heavyweight (which includes exception handling, for example), than simply adding support for new instructions should be good enough. I strongly encourage anyone willing to contribute, and maybe - given enough spare time - I will push some additional SSE instructions as PoC of how to add support for them.

You probably already saw that we support even some AES and AVX instructions, so adding support is quite trivial. Of course, some instructions are a pain to emulate, but where possible, intrinsics can be used for smaller, high performance code.

**Comment by JustasMasiulis** [comment]
> This sounds intriguing. Have you been able to compile bdshemu with Emscripten? I'm curious what changes were necessary for that to work.

Can't say for sure because I used cmake ([please don't judge this monstrosity](https://gist.github.com/JustasMasiulis/7a92d21f86161bd37e7d2a067be0708a)) to compile it using just LLVM+wasm-opt and it was quite a while ago. However I'd imagine using emscriptem should be a much better experience than what I did in the first place.

Did this under the assumptions that emscriptem would add extra overhead with its CRT/heap/extension implementations and that using cmake to compile the code without it would be as easy as compiling normal executable. Can for sure say that the second part was incorrect.

> You probably already saw that we support even some AES and AVX instructions, so adding support is quite trivial.

Now that AES was mentioned, I remembered that it was actually probably the only thing that needed to be removed, since there is no way to use the intrinsics in WASM and there wasn't a macro to disable their use.

> I strongly encourage anyone willing to contribute, and maybe - given enough spare time - I will push some additional SSE instructions as PoC of how to add support for them.

That would be great. In addition I think a few more lines in the docs/readme about the internals/how to contribute would go a long way.

**Comment by JustasMasiulis** [comment]
On an unrelated note, since I'm not sure whether this is worth its own issue, can I ask why does `RipAccess` never have the `ND_ACCESS_COND_WRITE` flag? Because the [docs suggesting](https://bddisasm.readthedocs.io/en/latest/#determining-if-the-instruction-is-a-branch-of-any-kind) to check for `ND_ACCESS_ANY_WRITE` to figure out whether it's any kind of branch led me to assume conditional jump instructions would have `ND_ACCESS_COND_WRITE` flag set. Caused me quite a bit of confusion and made me think my bindings were broken for some time.

**Comment by ianichitei** [comment]
> Can't say for sure because I used cmake ([please don't judge this monstrosity](https://gist.github.com/JustasMasiulis/7a92d21f86161bd37e7d2a067be0708a)) to compile it using just LLVM+wasm-opt and it was quite a while ago. However I'd imagine using emscriptem should be a much better experience than what I did in the first place.

That looks simple enough (if you want to see a monstrosity look at the older CMakeLists.txt I've written for this project 😆). I briefly toyed with emcc trying to compile a simple example that decoded a NOP with bddisasm and I was 50% successful: the page never finished loading when I built it with `WASM=1`, but worked well without it. I never did anything beyond that, I was mostly curious if it would work.

> Now that AES was mentioned, I remembered that it was actually probably the only thing that needed to be removed, since there is no way to use the intrinsics in WASM and there wasn't a macro to disable their use.

I'll add a check to ensure that the compiler supports AES before passing `-maes` to it and a cmake option to completely disable it. I try to consider build issues like this one a bug so I will gladly try to fix them. I don't know if I'll have the time to do it this weekend though.

**Comment by JustasMasiulis** [comment]
> That looks simple enough

Simple, yet took me half a day to sift through all the problems with cmake, windows and llvm 😩 

> Why was -fms-extensions needed? Is bddisasm behaving like it is compiled by MSVC?

I think I just copied a bunch of flags from another project and it's not necessary. Seems to compile fine with just `--target=wasm32 -flto`

> I don't know if I'll have the time to do it this weekend though.

I have absolutely no urgency for this and I doubt many other people are having issues like this so take your time 👍 

> I'll add a check to ensure that the compiler supports AES before passing -maes to it and a cmake option to completely disable it

LLVM didn't seem to care about the `-maes` flag. Only `#include <immintrin.h>` and the intrinsics themselves caused the errors.

On the topic of other minor things you could call "build" issues:

* With `BDDISASM_NO_FORMAT` defined I had quite a few warnings about the unused formatting string arrays (`gReg8Bit` ...  `gEmbeddedRounding`).
* The classic `#ifdef __cplusplus / extern "C" { / #endif` wrapping would be very much appreciated for use from C++

**Comment by vlutas** [comment]
> On an unrelated note, since I'm not sure whether this is worth its own issue, can I ask why does `RipAccess` never have the `ND_ACCESS_COND_WRITE` flag? Because the [docs suggesting](https://bddisasm.readthedocs.io/en/latest/#determining-if-the-instruction-is-a-branch-of-any-kind) to check for `ND_ACCESS_ANY_WRITE` to figure out whether it's any kind of branch led me to assume conditional jump instructions would have `ND_ACCESS_COND_WRITE` flag set. Caused me quite a bit of confusion and made me think my bindings were broken for some time.

No particular reason at all, other than "I've never ran into this issue before" and "Xed does it too". Looking at the pseudocode of the  conditional branches in Intel SDM, if the condition is not met, nothing is being done. Knowing a few microarchitectural details myself, not-taken branches will fall-through just like any regular instruction (ofc, ignoring uarch details such as branch prediction, faults or ucode assists that may still divert the execution flow).
TL;DR: It makes perfect sense for the RIP operand of conditional branches to be conditional write, and I'll do this change.

> LLVM didn't seem to care about the -maes flag. Only #include <immintrin.h> and the intrinsics themselves caused the errors.

wmmintrin.h should work for LLVM. We'll do the required modifications to be able to compile the AES intrinsics using LLVM too.

> On the topic of other minor things you could call "build" issues:

No biggie, can be fixed easily.

Thanks again for taking interest in our little toy!

**Comment by vlutas** [comment]
Hello!

If you check the latest commit, you'll see some changes, as discussed:

- RIP operands are conditional read/write for conditional branches (including JrCXZ, LOOP, XBEGIN) (starting with bddisasm 1.32.3);
- Added extern "C" to the functions declarations in bddisasm.h/bdshemu.h, so it plays nicely with C++;
- Fixed warning when building with BDDISASM_NO_FORMAT due to the string constants being unused;
- Include wmmintrin.h when building with clang;

If there are any other issues, please raise them. Otherwise, we'll close this one.

I also invite you to join our Slack channel - kvm-vmi [dot] slack [dot] com, where we have a dedicated bddisasm discussions channel, and we can exchange more ideas regarding how to make bdshemu more appealing for general purpose emulation.
